sycl/include/sycl/ext/intel/experimental/esimd/math.hpp

```diff
@@ -1727,6 +1727,91 @@ __ESIMD_API __ESIMD_NS::simd<T, N> dpasw2(
 }
 /// @} sycl_esimd_systolic_array_api
 
+/// @addtogroup sycl_esimd_logical
+/// @{
+
+/// Performs binary function computation with three vector operands.
+/// @tparam FuncControl boolean function control expressed with bfn_t
+/// enum values. Encodes all possible 256 functions with three boolean
+/// operands in an unsigned char.
+/// @tparam T type of the input vector element.
+/// @tparam N size of the input vector.
+/// @param s0 First boolean function argument.
+/// @param s1 Second boolean function argument.
+/// @param s2 Third boolean function argument.
+///
+/// Example: d = bfn<~bfn_x & ~bfn_y & ~bfn_z>(s0, s1, s2);
+
+enum class bfn_t : uint8_t { x = 0xAA, y = 0xCC, z = 0xF0 };
+
+/// The first argument of the boolean function.
+constexpr bfn_t bfn_x = bfn_t::x;
+/// The second argument of the boolean function.
+constexpr bfn_t bfn_y = bfn_t::y;
+/// The third argument of the boolean function.
+constexpr bfn_t bfn_z = bfn_t::z;
+
+static constexpr bfn_t operator~(bfn_t x) {
+  uint8_t val = static_cast<uint8_t>(x);
+  uint8_t res = ~val;
+  return static_cast<bfn_t>(res);
+}
+
+static constexpr bfn_t operator|(bfn_t x, bfn_t y) {
+  uint8_t arg0 = static_cast<uint8_t>(x);
+  uint8_t arg1 = static_cast<uint8_t>(y);
+  uint8_t res = arg0 | arg1;
+  return static_cast<bfn_t>(res);
+}
+
+static constexpr bfn_t operator&(bfn_t x, bfn_t y) {
+  uint8_t arg0 = static_cast<uint8_t>(x);
+  uint8_t arg1 = static_cast<uint8_t>(y);
+  uint8_t res = arg0 & arg1;
+  return static_cast<bfn_t>(res);
+}
+
+static constexpr bfn_t operator^(bfn_t x, bfn_t y) {
+  uint8_t arg0 = static_cast<uint8_t>(x);
+  uint8_t arg1 = static_cast<uint8_t>(y);
+  uint8_t res = arg0 ^ arg1;
+  return static_cast<bfn_t>(res);
+}
+
+template <bfn_t FuncControl, typename T, int N>
+__ESIMD_API std::enable_if_t<std::is_integral_v<T> &&
+                                 (sizeof(T) == 2 || sizeof(T) == 4),
```

Comment:
just for my own interest where does the 2/4 byte requirement come from? does the underlying hard instruction only support these?

Reply:
for int64_t it is trivial for us to support it by using bitcast to int32_t elements.
For int8_t it should also be possible to support by
a) if N - even: using bitcast to int16_t, bfn, bitcast to int16_t
b) if N - odd: extending with 1 element (to N+1 = even), bitcast to int16_t, bfn, extract N elements.

Reply:
> just for my own interest where does the 2/4 byte requirement come from? does the underlying hardware instruction only support these?

Yes, it comes from HW.

Reply:
> for int64_t it is trivial for us to support it by using bitcast to int32_t elements. For int8_t it should also be possible to support by a) if N - even: using bitcast to int16_t, bfn, bitcast to int16_t b) if N - odd: extending with 1 element (to N+1 = even), bitcast to int16_t, bfn, extract N elements.

Done.